### DISCUSSION TITLE
@oneOf Directive Support (help)

### DISCUSSION THREAD

**Discussion by maaft** [99designs/gqlgen#2280] [Ideas]
### What happened?
`@oneOf` is a commonly used directive on input-types. https://github.com/graphql/graphql-spec/pull/825

Using this directive, it is possible to validate that one (and only one) field from an input object is set, e.g:

```
input UserBy @oneOf {
  id: ID
  email: String
  username: String
  registrationNumber: Int
}

input UserInput {
  by: UserBy!
}

type Query {
  user(input: UserInput!): User
}
```
  

### What did you expect?

It would be great, if gqlgen could support and validate this directive.

I already tried to implement it by myself, but I'm confused by the directive function syntax:

```golang
c.Directives.OneOf = func(ctx context.Context, obj interface{}, next graphql.Resolver) (res interface{}, err error) {

		return next(ctx)
	}
```

In theory, I need to iterate through all keys of `obj` and check if only one of them is `!=nil`.

**Problems**:
 - This function seems to be called multiple times: First time for the `UserInput` type, then a 2nd time for the actual `UserBy` type on which the `@oneOf` directive is used.
 - I'm only interested in the `UserBy` input type
 
**Questions**:
 - How can I make sure to receive only `UserBy` types in my directive callback?
 - The solution should be **generic**. I **don't want to do type assertions** (like `_, ok := obj.(model.UserBy)`. The `@oneOf` directive will be used on multiple input times. I can't add a new type-assertion to the directive everytime a new input types uses that directive.

**Comment by maaft** [comment]
@frederikhors why move to discussion? This could be a bug IMHO

In my opinion, the directive callback should only be called with `UserBy` object, not with `UserInput` object first

**Reply by maaft** [reply]
I reworded it as an issue: https://github.com/99designs/gqlgen/issues/2281

**Reply by frederikhors** [reply]
Sorry to forgot a comment for you. You're right maybe it's a bug and I was wrong because I better understand it now in the reworded version #2281.
